### PR TITLE
[core] Drop `useGridContainerProps`

### LINKED ISSUE #1775: [Datagrid] The Grid flickers as it continuously resizes itself
Issue can be triggered after getting the resolution just right for the condition to take place.
In result the grid can be seen extending and shrinking horizontally non stop. After inspection of the source we can see JavaScript realigning width and min-width on a number of DOM Elements.

Experienced on both 4.0.0-alpha.28 and 4.0.0-alpha.29 versions and possibly on others.
Sometimes difficult to replicate, other times happens all the time. To try to replicate make window smaller and resize until you get scrollbar to show. It will likely disappear shortly after resizing. Now change the zoom level in and out - at some point the issue will occur. If it does not resize the window to a different size and try the steps again.

Any help, suggestions are really appreciated.

Attached GIF demonstrating the issue.

CPT2105271230-2181x702

### PR DESCRIPTION
**Pull request by flaviendelangle**
Rebased upon #3074 

Closes #1775 

## Changes

- [x] New method `apiRef.current.getRootDimensions` to get the current dimensions of the grid. Returns the following object

```ts
export interface GridDimensions {
  viewportOuterSize: ElementSize;
  viewportInnerSize: ElementSize;
  hasScrollX: boolean;
  hasScrollY: boolean;
}
```

- [x] New event  `viewportInnerSizeChange` to re-compute the column flex widths and the autoPageSize

- [x] New method `unstable_getViewportPageSize`

- [x] Remove `state.containerSizes`,  `state.viewportSizes` and `state.scrollBar`

## Render count

I counted the amount of rerenders in `GridBody` when mounting a `DataGrid` without any further interaction.
Those renders can be synchronous and do not mean that React is actually committing anything to the DOM.

### Async data fetching

```tsx
const { data } = useDemoData({
    dataSet: 'Commodity',
    rowLength: 10,
    maxColumns: 6,
});

return (
    <div style={{ height: 400, width: '100%' }}>
        <DataGrid {...data} autoPageSize />
    </div>
);
```

Before: 12 renders
After: 5 renders

### Sync data fetching with `props.autoPageSize`

```tsx
const data = React.useMemo(() => getData(10, 10), [])

return (
    <div style={{ height: 400, width: '100%' }}>
        <DataGrid {...data} autoPageSize />
    </div>
);
```

Before: 7 renders
After: 2 renders

### Sync data fetching without `props.autoPageSize`

```tsx
const data = React.useMemo(() => getData(10, 10), [])

return (
    <div style={{ height: 400, width: '100%' }}>
        <DataGrid {...data} />
    </div>
);
```

Before: 7 renders
After: 2 renders

## Why add this logic in `useGridDimensions` (ex `useGridResizeContainer`)

If we put this logic in `GridVirtualScroller`, then during the 1st render of the Rows and the Header, we don't have the dimensions yet. Which cuases flickering on the borders.

### CODE REVIEW

File: packages/grid/_modules_/grid/hooks/features/dimensions/useGridDimensions.ts
Comment (DanailH) [inline packages/grid/_modules_/grid/hooks/features/dimensions/useGridDimensions.ts:148]: Can we change this to `getGridDimentions` because it may not always be clear to what element do the dimensions belong?

File: packages/grid/_modules_/grid/components/GridVirtualScroller.tsx
Comment (DanailH) [inline packages/grid/_modules_/grid/components/GridVirtualScroller.tsx:9]: This should be a breaking change, no? The `gridContainerSizesSelector` is removed.

File: packages/grid/_modules_/grid/hooks/features/dimensions/useGridDimensions.ts
Comment (flaviendelangle) [reply packages/grid/_modules_/grid/hooks/features/dimensions/useGridDimensions.ts:148]: https://github.com/mui-org/material-ui-x/pull/3002

Once this one is merged it will be private so no breaking change :)

File: packages/grid/_modules_/grid/components/GridVirtualScroller.tsx
Comment (flaviendelangle) [reply packages/grid/_modules_/grid/components/GridVirtualScroller.tsx:9]: https://github.com/mui-org/material-ui-x/pull/3002

Once this one is merged it will be private so no breaking change :)

File: packages/grid/x-grid/src/tests/rows.DataGridPro.test.tsx
Comment (DanailH) [inline packages/grid/x-grid/src/tests/rows.DataGridPro.test.tsx:387]: I guess it's like this because it's a WIP.

File: packages/grid/_modules_/grid/hooks/features/dimensions/useGridDimensions.ts
Comment (flaviendelangle) [reply packages/grid/_modules_/grid/hooks/features/dimensions/useGridDimensions.ts]: Done

File: packages/grid/x-grid/src/tests/rows.DataGridPro.test.tsx
Comment (flaviendelangle) [reply packages/grid/x-grid/src/tests/rows.DataGridPro.test.tsx:387]: I would like to see with @m4theushw on this one.
I don't know how to check if the virtualization is on or off.
Since the new virtualization, these test do not check the right think, they check a state that was triggering the old virtualization but that is not used anymore.

**Review by m4theushw — COMMENTED** [review]
I would only be careful to not lack information into the dimensions object which is only used by one hook. This makes complex to change a value in the future as modifying it may cause side effects in other hooks.

File: packages/grid/_modules_/grid/hooks/features/infiniteLoader/useGridInfiniteLoader.ts
Comment (flaviendelangle) [reply packages/grid/_modules_/grid/hooks/features/infiniteLoader/useGridInfiniteLoader.ts:55]: Maybe the naming are to be improved.
In the previous implementation I think their were inconsistencies in what "virtual" meant.

In this PR, `rowsInViewportCount` is the amount of rows that are displayed on the screen and `virtualRowsCount` the amount of rows in the current page.
You can have `virtualRowsCount: 100000` and `rowsInViewportCount: 12`.


`rowsInViewportCount` is used for the `rowsScrollEnd` event and is the equivalent of the old `viewportPageSize`.
I tried to avoid any breaking change to be able to merge this one after v5

```ts
export interface GridRowScrollEndParams {
  /**
   * The number of rows that fit in the viewport.
   */
  viewportPageSize: number;
  /**
   * The number of rows allocated for the rendered zone.
   */
  virtualRowsCount: number;
  /**
   * The grid visible columns.
   */
  visibleColumns: GridColumns;
}
```

Comment (flaviendelangle) [reply packages/grid/_modules_/grid/hooks/features/infiniteLoader/useGridInfiniteLoader.ts:55]: I named it `rowsInViewportCount` and not `viewportPageSize` because it can be less than what would fit in the viewport.
If you have enough space on your scree to display 20 rows but you give 3, `rowsInViewportCount` will be equal to 3 not 20.

File: packages/grid/_modules_/grid/hooks/features/dimensions/gridDimensionsApi.ts
Comment (flaviendelangle) [reply packages/grid/_modules_/grid/hooks/features/dimensions/gridDimensionsApi.ts:23]: Currently in the Grid, `visible` = `post filter but not paginated`
It's poorly named as we discussed in an issue, but I would favor consistency and avoid naming `visibleXXX` anything that is paginated until we clear the naming in a future major.

Comment (flaviendelangle) [reply packages/grid/_modules_/grid/hooks/features/dimensions/gridDimensionsApi.ts:23]: `paginatedRowCount` ? `currentPageRowCount` ?

File: packages/grid/_modules_/grid/components/GridAutoSizer.tsx
Comment (flaviendelangle) [reply packages/grid/_modules_/grid/components/GridAutoSizer.tsx:100]: In `GridBody` we have

```
disableHeight={rootProps.autoHeight}
```

So when `autoHeight: true`, the event is never fired.
But we still need it to be to update the ref dimensions.

File: packages/grid/_modules_/grid/hooks/features/dimensions/gridDimensionsApi.ts
Comment (flaviendelangle) [reply packages/grid/_modules_/grid/hooks/features/dimensions/gridDimensionsApi.ts:38]: https://github.com/mui-org/material-ui-x/pull/3007#discussion_r740921219

As you want

File: packages/grid/_modules_/grid/hooks/features/dimensions/useGridDimensions.ts
Comment (flaviendelangle) [reply packages/grid/_modules_/grid/hooks/features/dimensions/useGridDimensions.ts:79]: The init is not needed since we always set a value later

File: packages/grid/_modules_/grid/hooks/features/keyboard/useGridKeyboardNavigation.ts
Comment (flaviendelangle) [reply packages/grid/_modules_/grid/hooks/features/keyboard/useGridKeyboardNavigation.ts:188]: Oh that's a bug, it should go down / up by the amount of rows in the viewport, not go to the end of the page.
I'll fix it

File: packages/grid/x-grid/src/tests/rows.DataGridPro.test.tsx
Comment (flaviendelangle) [reply packages/grid/x-grid/src/tests/rows.DataGridPro.test.tsx:387]: Removed

File: packages/grid/x-grid/src/tests/rows.DataGridPro.test.tsx
Comment (flaviendelangle) [reply packages/grid/x-grid/src/tests/rows.DataGridPro.test.tsx]: Done

File: packages/grid/_modules_/grid/hooks/features/dimensions/useGridDimensions.ts
Comment (m4theushw) [reply packages/grid/_modules_/grid/hooks/features/dimensions/useGridDimensions.ts]: Maybe `getRootDimensions` or even `getContainerDimensions`? We never used "grid" in an API method.

File: packages/grid/_modules_/grid/hooks/features/keyboard/useGridKeyboardNavigation.ts
Comment (flaviendelangle) [reply packages/grid/_modules_/grid/hooks/features/keyboard/useGridKeyboardNavigation.ts]: Fixed

File: packages/grid/_modules_/grid/hooks/features/infiniteLoader/useGridInfiniteLoader.ts
Comment (flaviendelangle) [reply packages/grid/_modules_/grid/hooks/features/infiniteLoader/useGridInfiniteLoader.ts:56]: If we want to release v5 without waiting for this PR, I would delay the renaming.
If we want to merge this one before, then I am open to suggestion to unify between this callback and the `GridDimensions` interface.

File: packages/grid/_modules_/grid/hooks/features/dimensions/useGridDimensions.ts
Comment (flaviendelangle) [reply packages/grid/_modules_/grid/hooks/features/dimensions/useGridDimensions.ts]: Let's go for `getRootDimensions`

File: packages/grid/_modules_/grid/hooks/features/dimensions/gridDimensionsApi.ts
Comment (m4theushw) [reply packages/grid/_modules_/grid/hooks/features/dimensions/gridDimensionsApi.ts:23]: Anything mentioning "pagination" won't make sense if pagination is disabled. `renderedRowCount`?

Comment (flaviendelangle) [reply packages/grid/_modules_/grid/hooks/features/dimensions/gridDimensionsApi.ts:23]: If virtualization is on, then not all rows listed in `virtualRowsCount` are rendered in the DOM.

The explicit name would be something like `amountOfRowsGivenToTheVirtualEngine` :smile: 

If pagination is on it's the amount of rows on the current page
If pagination is off it's the amount of visible rows

Comment (flaviendelangle) [reply packages/grid/_modules_/grid/hooks/features/dimensions/gridDimensionsApi.ts:23]: I understood the `virtual` in `virtualRowCount` as in `what we give to the virtualization engine, which does whatever he wants with it`.
So for me it made sense.

But we can probably find a better word to describe this quantity.

Comment (m4theushw) [reply packages/grid/_modules_/grid/hooks/features/dimensions/gridDimensionsApi.ts:23]: > The explicit name would be something like amountOfRowsGivenToTheVirtualEngine 😄

Also known as the number of visible rows that can be visible by scrolling all the way, so `visibleRowCount`. We could start doing the right thing here. `renderedRowCount` is also wrong as the virtualization doesn't render everything.

Comment (flaviendelangle) [reply packages/grid/_modules_/grid/hooks/features/dimensions/gridDimensionsApi.ts:23]: `scrollableRowCount` ?

File: packages/grid/_modules_/grid/hooks/features/keyboard/useGridKeyboardNavigation.ts
Comment (flaviendelangle) [reply packages/grid/_modules_/grid/hooks/features/keyboard/useGridKeyboardNavigation.ts]: I'm not so sure anymore, I'll check correctly tomorrow :+1:

Comment (flaviendelangle) [reply packages/grid/_modules_/grid/hooks/features/keyboard/useGridKeyboardNavigation.ts]: OK I'd like your feedback on this one.
If you take an actual demo, focus a cell on the 1st row and press space or page down, is goes down by the amount of rows you see on your screen (~12 in the demo I link below)
https://mui.com/components/data-grid/rows/#row-height

But we have this test `Space only should go to the bottom of the page` who wants it to go to the very bottom.

For me there was something wrong about the `containerSizes.viewportPageSize` value.
It is computed as follow:
```ts
const viewportPageSize =
  scrollBarState.hasScrollY || rowsCount < viewportFitHeightSize
    ? rowsCount
    : viewportFitHeightSize;
```
And I don't understand why when their is a scroll Y we take the row count (which is the value returned by `getVirtualRowCount`, so the size of the current page if paginated and the total amount of rows if not).

But it was used in the params of `rowsScrollEnd` where the JSDoc clearly says `The number of rows that fit in the viewport.`
So for me the behavior was broken.

Can we agree that when we press `Space` or `Page Down`, we want to go down by the amount of rows you can see on the screen ? I think that it was the intended behavior and it is what I see on Google Sheet for instance.

Comment (flaviendelangle) [reply packages/grid/_modules_/grid/hooks/features/keyboard/useGridKeyboardNavigation.ts]: Fixed

File: packages/grid/_modules_/grid/components/GridAutoSizer.tsx
Comment (flaviendelangle) [reply packages/grid/_modules_/grid/components/GridAutoSizer.tsx:100]: I changed the implementation to avoid this modification

File: packages/grid/_modules_/grid/hooks/features/keyboard/useGridKeyboardNavigation.ts
Comment (flaviendelangle) [reply packages/grid/_modules_/grid/hooks/features/keyboard/useGridKeyboardNavigation.ts]: For me the old behavior was kind of buggy.
I fixed it and the test to match the docs.

File: packages/grid/_modules_/grid/hooks/features/columns/gridColumnsSelector.ts
Comment (m4theushw) [reply packages/grid/_modules_/grid/hooks/features/columns/gridColumnsSelector.ts:25]: It seems that #1775 is already fixed in this PR without this change. Now the column widths will still be rendered with a non-integer value but `positions` and `totalWidth` will be integers. I remember that the idea of allowing non-integers was to not leave any blank space due the difference between the container width and the sum of the columns.

File: packages/grid/_modules_/grid/hooks/features/dimensions/gridDimensionsApi.ts
Comment (m4theushw) [reply packages/grid/_modules_/grid/hooks/features/dimensions/gridDimensionsApi.ts:23]: > scrollableRowCount ?

The rows might not be scrollable. Let's be explicit here. This value is the number of rows rendered by the virtualization engine, so: `virtualScrollerRowCount`. I don't have more suggestions.

File: packages/grid/_modules_/grid/hooks/features/infiniteLoader/useGridInfiniteLoader.ts
Comment (m4theushw) [reply packages/grid/_modules_/grid/hooks/features/infiniteLoader/useGridInfiniteLoader.ts:55]: > I tried to avoid any breaking change to be able to merge this one after v5

My idea would be to calculate this value in `useGridInfiniteLoader`. If it's not used in some other place we keep its calculation where it's used. If other feature asks for this value, then move it to the dimensions object. I want to avoid leaking this values that were created for a single feature.

File: packages/grid/_modules_/grid/hooks/features/columns/gridColumnsSelector.ts
Comment (flaviendelangle) [reply packages/grid/_modules_/grid/hooks/features/columns/gridColumnsSelector.ts:25]: I'm sceptical. I had the bug with this PR before but as you said if I remove the `Math.round` it still works.

For me the problem with the rounding was when we started rounding several columns, because at the rend, the rounded total was several pixel off the real total. 
But with this `Math.round` it should not be a problem.

I removed the `Math.round`. We will need to stay vigilant on that one.

File: packages/grid/_modules_/grid/hooks/features/infiniteLoader/useGridInfiniteLoader.ts
Comment (flaviendelangle) [reply packages/grid/_modules_/grid/hooks/features/infiniteLoader/useGridInfiniteLoader.ts:55]: It is also used in the keyboard navigation for the `Page Up` / `Page Down`

File: packages/grid/_modules_/grid/hooks/features/dimensions/gridDimensionsApi.ts
Comment (flaviendelangle) [reply packages/grid/_modules_/grid/hooks/features/dimensions/gridDimensionsApi.ts:23]: I removed it from `getRootDimensions` as it can be easily computed from the outside and was only used directly for the scroll end event.

Comment (flaviendelangle) [reply packages/grid/_modules_/grid/hooks/features/dimensions/gridDimensionsApi.ts:23]: Updated

File: packages/grid/_modules_/grid/hooks/features/dimensions/useGridDimensions.ts
Comment (flaviendelangle) [reply packages/grid/_modules_/grid/hooks/features/dimensions/useGridDimensions.ts]: Done

File: packages/grid/_modules_/grid/hooks/features/dimensions/useGridDimensions.ts
Comment (flaviendelangle) [reply packages/grid/_modules_/grid/hooks/features/dimensions/useGridDimensions.ts]: Done, the height was needed for `useGridPageSize` after applying your comment

File: packages/grid/_modules_/grid/hooks/features/dimensions/useGridDimensions.ts
Comment (flaviendelangle) [reply packages/grid/_modules_/grid/hooks/features/dimensions/useGridDimensions.ts]: Done

File: packages/grid/_modules_/grid/hooks/features/dimensions/useGridDimensions.ts
Comment (flaviendelangle) [reply packages/grid/_modules_/grid/hooks/features/dimensions/useGridDimensions.ts]: Done

File: packages/grid/_modules_/grid/hooks/features/dimensions/useGridDimensions.ts
Comment (flaviendelangle) [reply packages/grid/_modules_/grid/hooks/features/dimensions/useGridDimensions.ts]: Done

File: packages/grid/_modules_/grid/hooks/features/keyboard/useGridKeyboardNavigation.ts
Comment (flaviendelangle) [inline packages/grid/_modules_/grid/hooks/features/keyboard/useGridKeyboardNavigation.ts:130]: This logic is copy pasted in a few places. Maybe it is worth adding a `apiRef.current.unstable_getViewportPageSize`

File: packages/grid/_modules_/grid/hooks/features/dimensions/gridDimensionsApi.ts
Comment (m4theushw) [inline packages/grid/_modules_/grid/hooks/features/dimensions/gridDimensionsApi.ts:12]: Description missing.

File: packages/grid/_modules_/grid/hooks/features/dimensions/gridDimensionsApi.ts
Comment (m4theushw) [inline packages/grid/_modules_/grid/hooks/features/dimensions/gridDimensionsApi.ts:30]:
```suggestion
   * @returns {number} The amount of rows visible in the viewport
   * @ignore - do not document.
```

File: test/regressions/TestViewer.js
Comment (m4theushw) [inline test/regressions/TestViewer.js:42]: Why this change?

**Review by m4theushw — APPROVED** [review]

**Review by m4theushw — DISMISSED** [review]
Actually I don't wanna approve yet. There're a few regressions pointed by argos.

The lighter horizontal scrollbar means that the overlay is on top of it. This means that users are trapped and can't scroll to the right. The vertical alignment of the text is also wrong.

image

From what I found, `horizontalScrollbarSize` begins as 0 and is never updated once the value is available.

https://github.com/mui-org/material-ui-x/blob/2409df34aa924b4dd0d2ef80d59f2812974a1171/packages/grid/_modules_/grid/components/containers/GridOverlay.tsx#L56

I believe fixing this also fixes the inconsistency in the infinite loading demo.

File: test/regressions/TestViewer.js
Comment (flaviendelangle) [reply test/regressions/TestViewer.js:42]: Otherwise, in the e2e test you fake the timers so you never apply the `setTimeout` of `debouncedResize` and the dimensions ref is never set.

Comment (flaviendelangle) [reply test/regressions/TestViewer.js:42]: To give more context:
In `useGridDimensions` (and before in `useGridContainerResize`) we have the following logic:

```tsx
const isTestEnvironment = process.env.NODE_ENV === 'test';

[...]

if (isTestEnvironment) {
  // We don't need to debounce the resize for tests.
  resize();
  return;
}

debounceResize();
```

Which solves allow to test post-resize result synchronously in all tests except the `e2e` / `regressions`
In `e2e` tests we have no test impacted to my knowledge.
In `regressions` it causes issues because we never set the dimensions ref and therefore never triggers behavior like the `autoPageSize`.

I spent a few hours trying to fix this. I can't get Webpack to use `NODE_ENV === "test"` because it wants to have its mode equal to the `NODE_ENV` and it only has `development` and `production` modes.
So I just removed the mock of the timer, I am now only mocking the initial time to have reliable snapshots but not how the time evolve.

Comment (m4theushw) [reply test/regressions/TestViewer.js:42]: >  I spent a few hours trying to fix this. I can't get Webpack to use NODE_ENV === "test" because it wants to have its mode equal to the NODE_ENV and it only has development and production modes.

It may be an useless question, but did you try `webpack.DefinePlugin`?

Comment (flaviendelangle) [reply test/regressions/TestViewer.js:42]: Yes
If you put `NODE_ENV: "test"` it complains that the `NODE_ENV` should always be equal to the `mode`

File: scripts/exportsSnapshot.json
Comment (flaviendelangle) [inline scripts/exportsSnapshot.json:117]: This state has been removed

File: scripts/exportsSnapshot.json
Comment (flaviendelangle) [inline scripts/exportsSnapshot.json:61]: This state has been removed

**Review by m4theushw — APPROVED** [review]

**Pull request merged**
